// cucumber/lib/pickle_runner/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.version = undefined;
exports.getBinaryLocalPath = getBinaryLocalPath;
exports.getBinaryRemoteUrl = getBinaryRemoteUrl;

var _path = require('path');

var _path2 = _interopRequireDefault(_path);

var _os = require('os');

var _os2 = _interopRequireDefault(_os);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var version = exports.version = 'v0.0.2';

function throwUnsupportedErrorMessage() {
  throw new Error('Unsupported operating system (' + process.platform + ') and system architecture (' + process.arch + ') combination. Please open an issue on cucumber-js.');
}

function getGoBinaryOperatingSystem() {
  switch (process.platform) {
    case 'darwin':
      return 'darwin';
    case 'freebsd':
      return 'freebsd';
    case 'linux':
      return 'linux';
    case 'win32':
      return 'windows';
    case 'openbsd':
      return 'openbsd';
    default:
      throwUnsupportedErrorMessage();
  }
}

function getGoBinaryArchitecture() {
  // Special ckeck on windows machines
  if (process.env.PROCESSOR_ARCHITECTURE === 'AMD64' || process.env.PROCESSOR_ARCHITEW6432 === 'AMD64') {
    return 'amd64';
  }
  switch (process.arch) {
    case 'arm':
      return 'arm';
    case 'ia32':
    case 'x32':
      return '386';
    case 'x64':
      return 'amd64';
    case 'mips':
      return 'mips';
    case 'mipsel':
      return 'mipsle';
    case 's390x':
      return 's390x';
    default:
      throwUnsupportedErrorMessage();
  }
}

function getBinaryLocalPath() {
  return _path2.default.join(_os2.default.homedir(), 'cucumber', 'pickle-runner', version);
}

function getBinaryRemoteUrl() {
  var urlPrefix = 'https://github.com/cucumber/cucumber-pickle-runner/releases/download';
  var os = getGoBinaryOperatingSystem();
  var arch = getGoBinaryArchitecture();
  var binaryName = 'cucumber-pickle-runner-' + os + '-' + arch;
  if (os === 'windows') {
    binaryName += '.exe';
  }
  return urlPrefix + '/' + version + '/' + binaryName;
}
