// cucumber/lib/cli/step_definition_snippet_builder/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _step_arguments = require('../../step_arguments');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var StepDefinitionSnippetBuilder = function () {
  function StepDefinitionSnippetBuilder(_ref) {
    var snippetSyntax = _ref.snippetSyntax;
    (0, _classCallCheck3.default)(this, StepDefinitionSnippetBuilder);

    this.snippetSyntax = snippetSyntax;
  }

  (0, _createClass3.default)(StepDefinitionSnippetBuilder, [{
    key: 'parameterTypeNamesToParameterNames',
    value: function parameterTypeNamesToParameterNames(parameterTypeNames) {
      var mapping = {};
      return parameterTypeNames.map(function (typeName) {
        if (!mapping[typeName]) {
          mapping[typeName] = 1;
        }
        var result = '' + typeName + mapping[typeName];
        mapping[typeName] += 1;
        return result;
      });
    }
  }, {
    key: 'build',
    value: function build(_ref2) {
      var _this = this;

      var generatedExpressions = _ref2.generatedExpressions,
          pickleArguments = _ref2.pickleArguments;

      var updatedGeneratedExpressions = generatedExpressions.map(function (_ref3) {
        var text = _ref3.text,
            parameterTypeNames = _ref3.parameterTypeNames;

        return {
          text: text,
          parameterNames: _this.parameterTypeNamesToParameterNames(parameterTypeNames)
        };
      });
      var comment = 'Write code here that turns the phrase above into concrete actions';
      var stepParameterNames = pickleArguments.map((0, _step_arguments.buildStepArgumentIterator)({
        dataTable: function dataTable() {
          return 'dataTable';
        },
        docString: function docString() {
          return 'docString';
        }
      }));
      return this.snippetSyntax.build({
        comment: comment,
        generatedExpressions: updatedGeneratedExpressions,
        stepParameterNames: stepParameterNames
      });
    }
  }]);
  return StepDefinitionSnippetBuilder;
}();

exports.default = StepDefinitionSnippetBuilder;
